### PR TITLE
Feature/sc 32779/search module route updates for clark refactor

### PR DESCRIPTION
**Pull request by Romerico234**
Removed and updated GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER route to point to search.service.ts from featured.service.ts and learning-object.service.ts

### CODE REVIEW

**Review by Cwagne17 — COMMENTED** [review]
Can you rebase to the branch you worked off of and then re-request.

File: src/app/core/learning-object-module/search/search.service.ts [hunk 1]
```diff
@@ -1,9 +1,94 @@
...
+  constructor(private http: HttpClient) {}
 
-  constructor() { }
+  /**
+   * Fetches Array of Learning Objects
+   *
+   * @returns {Promise<LearningObject[]>}
+   * @memberof LearningObjectService
+   */
+  getLearningObjects(
```
Comment (pzalep1) [inline src/app/core/learning-object-module/search/search.service.ts:22]: Same thing as the other PRs. If you change the class that the function is in all of the references will need to be updated. VS Code has some shortcuts that make it easy to find where the references are.

**Review by pzalep1 — COMMENTED** [review]
Looks good! Just need to update the references to the functions that you moved.

**Comment by Romerico234** [comment]
Working on fixing function calls

**Comment by Romerico234** [comment]
There was another search.service.ts in the admin directory so I removed it and moved its functions to search.service.ts in the  learning-object-module. And I updated the function calls.

File: src/app/core/feature-module/featured.service.ts [hunk 2]
```diff
@@ -148,34 +145,6 @@ export class FeaturedObjectsService {
...
-
-    return this.http
-      .get(route)
-      .pipe(catchError(this.handleError))
-      .toPromise()
-      .then((response: any) => {
-        return { learningObjects: response.objects, total: response.total };
-      });
-  }
-
```
Comment (pzalep1) [inline src/app/core/feature-module/featured.service.ts:178]: I believe this should still be located here or the admin dashboard will break.

Comment (Cwagne17) [reply src/app/core/feature-module/featured.service.ts:178]: This is actually using a learning-objects search route. I would update the component that uses it and it should be okay. I just think the name of the function should be updated to match how we refer to it in the backend.

File: src/app/core/learning-object-module/search/search.service.ts [hunk 3]
```diff
@@ -1,9 +1,100 @@
...
+        queryClone.standardOutcomes.length &&
+        typeof queryClone.standardOutcomes[0] !== 'string'
+      ) {
+        queryClone.standardOutcomes = (
+          queryClone.standardOutcomes as string[]
+        ).map((o) => o['id']);
+      }
+      const queryString = new URLSearchParams(queryClone).toString();
+      route =
+        SEARCH_ROUTES.GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER(queryString);
```
Comment (Cwagne17) [inline src/app/core/learning-object-module/search/search.service.ts:41]: can we rename the search route to match the name we use in the backend? `searchLearningObjects` and then condense these functions into one?

File: src/app/core/learning-object-module/search/search.service.ts [hunk 4]
```diff
@@ -1,9 +1,100 @@
...
+import { Query } from 'app/interfaces/query';
+import * as querystring from 'querystring';
+import { SEARCH_ROUTES } from 'app/core/learning-object-module/search/search.routes';
 
 @Injectable({
-  providedIn: 'root'
+  providedIn: 'root',
 })
 export class SearchService {
+  needsChange$: Subject<void> = new Subject();
```
Comment (Cwagne17) [inline src/app/core/learning-object-module/search/search.service.ts:15]: This route needs to be moved into the search module also. This comes from the learning object search module route handler in clark-service. 

https://github.com/Cyber4All/clark-service/blob/main/src/modules/learning-object-module/search/search.routehandler.ts

Screenshot 2024-09-16 at 10 09 34 AM

**Review by Cwagne17 — CHANGES_REQUESTED** [review]

**Comment by Romerico234** [comment]
I combined the two functions into a single function, searchLearningObjects. To distinguish between the functionalities of the original functions within this combined function, I added optional parameters that handle standard outcomes and the mapping to LearningObject instances. These optional parameters are set to true by default.

Additionally, I updated the function calls to this new function to specify the appropriate options. Specifically, for any component that previously used getNotPublicLearningObjects (which was only one), I set both options to false to replicate its original behavior.

File: src/app/admin/pages/learning-objects/learning-objects.component.ts [hunk 5]
```diff
@@ -18,11 +17,12 @@ import { ToastrOvenService } from '../../../shared/modules/toaster/notification.
 import { AuthService } from '../../../core/auth-module/auth.service';
 import { Collection } from '../../../core/collection-module/collections.service';
 import { UserService } from 'app/core/user-module/user.service';
+import { SearchService } from 'app/core/learning-object-module/search/search.service';
 @Component({
   selector: 'clark-learning-objects',
   templateUrl: './learning-objects.component.html',
   styleUrls: ['./learning-objects.component.scss'],
-  providers: [PublicLearningObjectService]
+  providers: []
```
Comment (johngonz0) [inline src/app/admin/pages/learning-objects/learning-objects.component.ts:25]: does there need to be anything in this providers array?

**Review by johngonz0 — COMMENTED** [review]
just a question

File: src/app/admin/pages/learning-objects/learning-objects.component.ts (hunk 5 above)
Comment (pzalep1) [reply src/app/admin/pages/learning-objects/learning-objects.component.ts:25]: Not with the new organizational structure we are doing with the services.

**Review by pzalep1 — APPROVED** [review]

File: src/app/core/learning-object-module/search/search.routes.ts [hunk 6]
```diff
@@ -1,8 +1,16 @@
...
+
+const needsChange$: Subject<void> = new Subject<void>();
 
 export const SEARCH_ROUTES = {
-  GET_PUBLIC_LEARNING_OBJECTS: `${environment.apiURL}/learning-objects`,
-  GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER(query: string) {
+  needsChange$,
+  REGISTER_CHANGE() {
+    this.needsChange$.next();
+  },
```
Comment (Cwagne17) [inline src/app/core/learning-object-module/search/search.routes.ts:10]: This should not be here, only routes to our APIs should be in the routes.ts files.

File: src/app/core/learning-object-module/search/search.routes.ts [hunk 7]
```diff
@@ -1,8 +1,16 @@
...
+
+  SEARCH_LEARNING_OBJECTS: `${environment.apiURL}/learning-objects`,
+  SEARCH_LEARNING_OBJECTS_WITH_FILTER(query: string) {
```
Comment (Cwagne17) [inline src/app/core/learning-object-module/search/search.routes.ts:13]: These can be combined.

File: src/app/core/learning-object-module/search/search.service.ts [hunk 8]
```diff
@@ -1,9 +1,90 @@
...
+   * @param {boolean} [options.mapToLearningObjectInstance=true] - Set false when you need to receive raw objects instead of LearningObject
+   * @returns {Promise<LearningObject[]>}
+   * @memberof LearningObjectService
+   * */
+
+  searchLearningObjects(
+    query?: Query,
+    options?: {
+      handleStandardOutcomes?: boolean;
+      mapToLearningObjectInstance?: boolean;
```
Comment (Cwagne17) [inline src/app/core/learning-object-module/search/search.service.ts:29]: When would we not want to map to a learning object class?

File: src/app/core/learning-object-module/search/search.service.ts [hunk 9]
```diff
@@ -1,9 +1,90 @@
...
+          : objects;
+
+        return { learningObjects, total };
+      });
+  }
+
+  registerChange() {
+    SEARCH_ROUTES.REGISTER_CHANGE();
+  }
+
```
Comment (Cwagne17) [inline src/app/core/learning-object-module/search/search.service.ts:80]: We are still missing getUserLearningObjects function.

**Review by Cwagne17 — CHANGES_REQUESTED** [review]

**Comment by Romerico234** [comment]
Waiting for Chris' PR to get merged then we'll see what to do with this PR.

File: src/app/core/learning-object-module/search/search.routes.ts (hunk 6 above)
Comment (Romerico234) [reply src/app/core/learning-object-module/search/search.routes.ts]: I'm a little confused with what you mean by this. Are you asking me to move this route into the search.routehandler.ts file in clark-service?

**Comment by Romerico234** [comment]
I moved registerChange route back to the search.service.ts in admin/core directory. I combined the function and variable in search.route.ts. Lastly, I pulled updates from clark-service-refactor.

**Review by Cwagne17 — APPROVED** [review]

**Review by stef359 — APPROVED** [review]

**Review by pzalep1 — APPROVED** [review]

**Review by johngonz0 — APPROVED** [review]

**Pull request merged**